File: agent/src/InfraControl.Agent.Core/Services/AgentRuntimeOrchestrator.cs
using InfraControl.Agent.Core.Constants;
using InfraControl.Agent.Core.Models;

namespace InfraControl.Agent.Core.Services;

public sealed class AgentRuntimeOrchestrator
{
    private readonly IAgentConfigurationStore _configurationStore;
    private readonly IAgentStatusStore _statusStore;
    private readonly IDeviceInventoryCollector _inventoryCollector;
    private readonly IAgentApiClient _apiClient;
    private readonly HeartbeatPayloadFactory _payloadFactory;
    private readonly BackoffPolicy _backoffPolicy;
    private int _consecutiveFailures;

    public AgentRuntimeOrchestrator(
        IAgentConfigurationStore configurationStore,
        IAgentStatusStore statusStore,
        IDeviceInventoryCollector inventoryCollector,
        IAgentApiClient apiClient,
        HeartbeatPayloadFactory payloadFactory,
        BackoffPolicy backoffPolicy)
    {
        _configurationStore = configurationStore;
        _statusStore = statusStore;
        _inventoryCollector = inventoryCollector;
        _apiClient = apiClient;
        _payloadFactory = payloadFactory;
        _backoffPolicy = backoffPolicy;
    }

    public async Task<TimeSpan> ExecuteOnceAsync(CancellationToken cancellationToken = default)
    {
        AgentConfiguration configuration = await _configurationStore.LoadAsync(cancellationToken);

        if (!configuration.HasServerUrl || (!configuration.HasAgentToken && !configuration.HasEnrollmentToken))
        {
            await SaveStatusAsync(AgentRuntimeStatusCode.NotConfigured, "Server URL and enrollment token are required before the agent can start.", cancellationToken);

            _consecutiveFailures = 0;
            return TimeSpan.FromSeconds(AgentConstants.DefaultHeartbeatIntervalSeconds);
        }

        DeviceSnapshot device = await _inventoryCollector.CaptureDeviceAsync(cancellationToken);

        if (!configuration.HasAgentToken)
        {
            await SaveStatusAsync(AgentRuntimeStatusCode.Enrolling, "Enrolling the device with InfraControl.", cancellationToken);

            EnrollOperationResult enrollResult = await _apiClient.EnrollAsync(configuration, _payloadFactory.CreateEnrollRequest(configuration, device), cancellationToken);

            if (!enrollResult.Success || enrollResult.Response is null)
            {
                return await HandleEnrollFailureAsync(enrollResult, cancellationToken);
            }

            configuration.AgentToken = enrollResult.Response.AgentToken;
            configuration.EnrollmentToken = null;
            configuration.HeartbeatIntervalSeconds = NormalizeInterval(enrollResult.Response.IntervalSeconds);
            await _configurationStore.SaveAsync(configuration, cancellationToken);
            await SaveStatusAsync(AgentRuntimeStatusCode.Enrolled, "Enrollment completed successfully.", cancellationToken);
            _consecutiveFailures = 0;
        }

        InventorySnapshot inventory = await _inventoryCollector.CaptureInventoryAsync(cancellationToken);
        HeartbeatRequest heartbeat = _payloadFactory.CreateHeartbeatRequest(device, inventory);
        HeartbeatOperationResult heartbeatResult = await _apiClient.SendHeartbeatAsync(configuration, heartbeat, cancellationToken);

        if (!heartbeatResult.Success || heartbeatResult.Response is null)
        {
            return await HandleHeartbeatFailureAsync(heartbeatResult, cancellationToken);
        }

        configuration.HeartbeatIntervalSeconds = NormalizeInterval(heartbeatResult.Response.NextIntervalSeconds);
        await _configurationStore.SaveAsync(configuration, cancellationToken);
        await SaveStatusAsync(AgentRuntimeStatusCode.HeartbeatHealthy, "Heartbeat sent successfully.", cancellationToken);

        _consecutiveFailures = 0;

        return TimeSpan.FromSeconds(configuration.HeartbeatIntervalSeconds);
    }

    private async Task<TimeSpan> HandleEnrollFailureAsync(EnrollOperationResult result, CancellationToken cancellationToken)
    {
        _consecutiveFailures++;

        AgentRuntimeStatusCode statusCode = result.FailureKind switch
        {
            AgentApiFailureKind.InvalidEnrollmentToken => AgentRuntimeStatusCode.InvalidEnrollmentToken,
            AgentApiFailureKind.ServerUnreachable => AgentRuntimeStatusCode.ServerUnreachable,
            _ => AgentRuntimeStatusCode.Error,
        };

        string message = result.Message ?? "Enrollment failed.";
        await SaveStatusAsync(statusCode, message, cancellationToken);

        return _backoffPolicy.GetDelay(_consecutiveFailures);
    }

    private async Task<TimeSpan> HandleHeartbeatFailureAsync(HeartbeatOperationResult result, CancellationToken cancellationToken)
    {
        _consecutiveFailures++;

        AgentRuntimeStatusCode statusCode = result.FailureKind switch
        {
            AgentApiFailureKind.ServerUnreachable => AgentRuntimeStatusCode.ServerUnreachable,
            AgentApiFailureKind.Unauthorized => AgentRuntimeStatusCode.HeartbeatDegraded,
            _ => AgentRuntimeStatusCode.Error,
        };

        string message = result.Message ?? "Heartbeat failed.";
        await SaveStatusAsync(statusCode, message, cancellationToken);

        return _backoffPolicy.GetDelay(_consecutiveFailures);
    }

    private async Task SaveStatusAsync(AgentRuntimeStatusCode statusCode, string message, CancellationToken cancellationToken)
    {
        AgentStatusSnapshot current = await _statusStore.LoadAsync(cancellationToken);

        current.StatusCode = statusCode;
        current.Message = message;

        await _statusStore.SaveAsync(current, cancellationToken);
    }

    private static int NormalizeInterval(int intervalSeconds) =>
        intervalSeconds > 0 ? intervalSeconds : AgentConstants.DefaultHeartbeatIntervalSeconds;
}

File: agent/src/services/agent-runtime-orchestrator.mjs
import {
  AgentApiFailureKind,
  AgentRuntimeStatusCode,
  DEFAULT_HEARTBEAT_INTERVAL_SECONDS,
} from '../constants.mjs';
import {
  hasAgentToken,
  hasEnrollmentToken,
  hasServerUrl,
} from './file-agent-configuration-store.mjs';

function normalizeInterval(intervalSeconds) {
  return intervalSeconds > 0 ? intervalSeconds : DEFAULT_HEARTBEAT_INTERVAL_SECONDS;
}

export function createAgentRuntimeOrchestrator({
  configurationStore,
  statusStore,
  inventoryCollector,
  apiClient,
  payloadFactory,
  backoffPolicy,
}) {
  let consecutiveFailures = 0;

  return {
    async executeOnce() {
      const configuration = await configurationStore.load();

      if (!hasServerUrl(configuration) || (!hasAgentToken(configuration) && !hasEnrollmentToken(configuration))) {
        await saveStatus(statusStore, AgentRuntimeStatusCode.NotConfigured, 'Server URL and enrollment token are required before the agent can start.');
        consecutiveFailures = 0;
        return DEFAULT_HEARTBEAT_INTERVAL_SECONDS;
      }

      const device = await inventoryCollector.captureDevice();

      if (!hasAgentToken(configuration)) {
        await saveStatus(statusStore, AgentRuntimeStatusCode.Enrolling, 'Enrolling the device with InfraControl.');

        const enrollResult = await apiClient.enroll(
          configuration,
          payloadFactory.createEnrollRequest(configuration, device),
        );

        if (!enrollResult.success || !enrollResult.response) {
          consecutiveFailures += 1;
          const statusCode =
            enrollResult.failureKind === AgentApiFailureKind.InvalidEnrollmentToken
              ? AgentRuntimeStatusCode.InvalidEnrollmentToken
              : enrollResult.failureKind === AgentApiFailureKind.ServerUnreachable
                ? AgentRuntimeStatusCode.ServerUnreachable
                : AgentRuntimeStatusCode.Error;

          await saveStatus(statusStore, statusCode, enrollResult.message ?? 'Enrollment failed.');
          return backoffPolicy.getDelaySeconds(consecutiveFailures);
        }

        configuration.agentToken = enrollResult.response.agentToken;
        configuration.enrollmentToken = null;
        configuration.heartbeatIntervalSeconds = normalizeInterval(enrollResult.response.intervalSeconds);
        await configurationStore.save(configuration);
        await saveStatus(statusStore, AgentRuntimeStatusCode.Enrolled, 'Enrollment completed successfully.');
        consecutiveFailures = 0;
      }

      const inventory = await inventoryCollector.captureInventory();
      const heartbeatResult = await apiClient.sendHeartbeat(
        configuration,
        payloadFactory.createHeartbeatRequest(device, inventory),
      );

      if (!heartbeatResult.success || !heartbeatResult.response) {
        consecutiveFailures += 1;
        const statusCode =
          heartbeatResult.failureKind === AgentApiFailureKind.ServerUnreachable
            ? AgentRuntimeStatusCode.ServerUnreachable
            : heartbeatResult.failureKind === AgentApiFailureKind.Unauthorized
              ? AgentRuntimeStatusCode.HeartbeatDegraded
              : AgentRuntimeStatusCode.Error;

        await saveStatus(statusStore, statusCode, heartbeatResult.message ?? 'Heartbeat failed.');
        return backoffPolicy.getDelaySeconds(consecutiveFailures);
      }

      configuration.heartbeatIntervalSeconds = normalizeInterval(heartbeatResult.response.nextIntervalSeconds);
      await configurationStore.save(configuration);
      await saveStatus(statusStore, AgentRuntimeStatusCode.HeartbeatHealthy, 'Heartbeat sent successfully.');
      consecutiveFailures = 0;

      return configuration.heartbeatIntervalSeconds;
    },
  };
}

async function saveStatus(statusStore, statusCode, message) {
  const current = await statusStore.load();
  current.statusCode = statusCode;
  current.message = message;
  await statusStore.save(current);
}
